static/js/reactions.js

```diff
@@ -501,32 +521,115 @@ export function set_clean_reactions(message) {
 
         message.clean_reactions.set(
             local_id,
-            make_clean_reaction({local_id, user_ids, ...reaction}),
+            make_clean_reaction({message, local_id, user_ids, ...reaction}),
         );
     }
 }
 
-function make_clean_reaction({local_id, user_ids, emoji_name, emoji_code, reaction_type}) {
+function make_clean_reaction({message, local_id, user_ids, emoji_name, emoji_code, reaction_type}) {
     const r = {
         local_id,
         user_ids,
         ...emoji.get_emoji_details_for_rendering({emoji_name, emoji_code, reaction_type}),
     };
 
-    update_user_fields(r);
+    update_user_fields(r, message);
+    r.vote_text = get_vote_text_in_clean_reactions(r.user_ids, message);
 
     r.emoji_alt_code = user_settings.emojiset === "text";
     r.is_realm_emoji = r.reaction_type === "realm_emoji" || r.reaction_type === "zulip_extra_emoji";
 
     return r;
 }
 
-export function update_user_fields(r) {
+export function update_user_fields(r, message) {
     r.count = r.user_ids.length;
+    r.vote_text = get_vote_text(r.user_ids, message);
     r.label = generate_title(r.emoji_name, r.user_ids);
     if (r.user_ids.includes(page_params.user_id)) {
         r.class = "message_reaction reacted";
     } else {
         r.class = "message_reaction";
     }
 }
+
+// to set vote text when message is being cleaned.
+// this is an alternative to the commented out 'if'
+// block in get_total_votes_on_message() for now.
+function get_vote_text_in_clean_reactions(user_list, message) {
+    const helpers = get_helpers;
+    if (!helpers().should_display_emoji_reaction_users()) {
+        return `${user_list.length}`;
+    }
+    const usernames = people.get_display_full_names(user_list);
+    const current_user_has_reacted = user_list.includes(page_params.user_id);
+
+    if (current_user_has_reacted) {
+        const current_user_index = user_list.indexOf(page_params.user_id);
+        usernames[current_user_index] = $t({
+            defaultMessage: "You",
+        });
+    }
+
+    if (message.reactions.length <= 3) {
+        const comma_separated_usernames = usernames.join(", ");
+        return comma_separated_usernames;
+    }
+
+    // returning length as a string instead of integer for consistency
+    return `${usernames.length}`;
+}
```

Comment:
I think we want to figure out a way to delete this function, since it's a near duplicate and those tend to be for maintainability. Would a valid approach be to just not set `vote_text` at all in the `get_clean_reactions`, and just have `update_vote_text_on_message` and other code paths that actually render a reaction be responsible for setting that field?

Reply:
@timabbott `update_vote_text_on_message` only works for messages with cleaned reactions. In [this](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Correct.20way.20to.20get.20total.20votes.20on.20a.20message/near/1332590) thread, @showell suggested that I make these helper functions dumb and rely on previous functions to clean the reactions (I was earlier handling both clean and unclean cases in this function).
So, the `vote_text`(like other user fields) _has_ to be set in `get_clean_reactions`.

I agree that we need to do something about the duplicate function problem. Let me know what you think about this approach: 
```
function get_vote_text_in_clean_reactions(user_list, message) {
    if (!user_settings.display_emoji_reaction_users || message.reactions.length > 3) {
        return `${user_list.length}`;
    }
    return comma_separated_usernames(user_list);
}

export function get_vote_text(user_list, message) {
    if (!user_settings.display_emoji_reaction_users || get_total_votes_on_message(message) > 3) {
        return `${user_list.length}`;
    }
    return comma_separated_usernames(user_list);
}

function comma_separated_usernames(user_list) {
    const usernames = people.get_display_full_names(user_list);
    const current_user_has_reacted = user_list.includes(page_params.user_id);

    if (current_user_has_reacted) {
        const current_user_index = user_list.indexOf(page_params.user_id);
        usernames[current_user_index] = $t({
            defaultMessage: "You",
        });
    }
    const comma_separated_usernames = usernames.join(", ");
    return comma_separated_usernames;
}
```

Reply:
I still don't understand why one can't just have a rule that `get_vote_text` is only called on cleaned reactions, though. The linked conversation with Steve doesn't explain what code path prevents the model I suggested from working. Namely, that we only call `get_vote_text` when we're about to render the vote text to the DOM (and thus we can assume we already have cleaned reactions). 

If there is a good reason for this structure, we will want to document the reason in a comment, because as it is, the first thing I'd do when working with this code is try to refactor it.

Reply:
Is the motivation for this odd structure node testing considerations? I think it'd be simple enough to just test `get_vote_text` itself...

Reply:
@timabbott The `vote_text` property is analogous to the `count` property which it is replacing. The way I went about implementing this was replacing the `count` with `vote_text` everywhere. So just like the `count`, the `vote_text` also has to be set inside `make_clean_reactions`. Looking at it from this perspective, I can't think of a way to avoid setting the `vote_text` in `make_clean_reactions`.

The `vote_text` is actually already being set in `update_user_fields` which is called in `make_clean_reactions`. The line of code setting it again is **overwriting** it, since unlike `count` (which was easily set as `count = user_list.length`), the value of `vote_text` has to be retrieved retrieved differently for messages with cleaned and uncleaned reactions.

Reply:
Can you also clarify the model you had in mind in your previous [comment](https://github.com/zulip/zulip/pull/21013#discussion_r856623060)? I just want to make sure if I understood it correctly.
Thanks!

Reply:
What I meant is that we'd only call `get_vote_text` just before we render HTML that would contain the vote text -- not trying to precompute/cache that on `cleaned_reactons` objects.

Reply:
@timabbott After going through the code this is what I have gathered -> 

> I still don't understand why one can't just have a rule that `get_vote_text` is only called on cleaned reactions, though.

This is how the `vote_text` (and previously `count`) is rendered when a user reloads or loads the page for the first time.
```html
<div class="message_reaction_count">{{this.vote_text}}</div>
```
So, in order to correctly populate this `vote_text` initially, it _needs_ to be attached to the `clean_reaction` object. This means we have to set `vote_text` in `make_clean_reactions()` using `get_vote_text`, if the existing workflow is to be followed.

So the model you suggested wouldn't work because the code looks at the `vote_text` value in `clean_reactions` when initially, the data is being populated from the database.